xbmc/utils/EGLUtils.cpp

```diff
@@ -75,76 +77,92 @@ CEGLContextUtils::~CEGLContextUtils()
   Destroy();
 }
 
-bool CEGLContextUtils::CreateDisplay(EGLDisplay display,
-                                     EGLint renderable_type,
-                                     EGLint rendering_api)
+bool CEGLContextUtils::CreateDisplay(EGLenum platform, void* nativeDisplay, EGLint renderableType, EGLint renderingApi)
 {
-  EGLint neglconfigs = 0;
-  int major, minor;
-
-  EGLint surface_type = EGL_WINDOW_BIT;
-  // for the non-trivial dirty region modes, we need the EGL buffer to be preserved across updates
-  if (g_advancedSettings.m_guiAlgorithmDirtyRegions == DIRTYREGION_SOLVER_COST_REDUCTION ||
-      g_advancedSettings.m_guiAlgorithmDirtyRegions == DIRTYREGION_SOLVER_UNION)
-    surface_type |= EGL_SWAP_BEHAVIOR_PRESERVED_BIT;
-
-  EGLint attribs[] =
+  if (m_eglDisplay != EGL_NO_DISPLAY)
   {
-    EGL_RED_SIZE,        8,
-    EGL_GREEN_SIZE,      8,
-    EGL_BLUE_SIZE,       8,
-    EGL_ALPHA_SIZE,      8,
-    EGL_DEPTH_SIZE,     16,
-    EGL_STENCIL_SIZE,    0,
-    EGL_SAMPLE_BUFFERS,  0,
-    EGL_SAMPLES,         0,
-    EGL_SURFACE_TYPE,    surface_type,
-    EGL_RENDERABLE_TYPE, renderable_type,
-    EGL_NONE
-  };
+    throw std::logic_error("Do not call CreateDisplay when display has already been created");
+  }
 
-#if defined(EGL_EXT_platform_base) && defined(EGL_KHR_platform_gbm) && defined(HAVE_GBM)
-  if (m_eglDisplay == EGL_NO_DISPLAY &&
-      CEGLUtils::HasExtension(EGL_NO_DISPLAY, "EGL_EXT_platform_base") &&
-      CEGLUtils::HasExtension(EGL_NO_DISPLAY, "EGL_KHR_platform_gbm"))
+#if defined(EGL_EXT_platform_base)
+  if (platform != EGL_NONE && CEGLUtils::HasClientExtension("EGL_EXT_platform_base"))
   {
-    PFNEGLGETPLATFORMDISPLAYEXTPROC getPlatformDisplayEXT = (PFNEGLGETPLATFORMDISPLAYEXTPROC)eglGetProcAddress("eglGetPlatformDisplayEXT");
-    if (getPlatformDisplayEXT)
+    // Theoretically it is possible to use eglGetDisplay() and eglCreateWindowSurface,
+    // but then the EGL library basically has to guess which platform we want
+    // if it supports multiple which is usually the case -
+    // it's better and safer to make it explicit
+
+    auto getPlatformDisplayEXT = CEGLUtils::GetRequiredProcAddress<PFNEGLGETPLATFORMDISPLAYEXTPROC>("eglGetPlatformDisplayEXT");
+    m_eglDisplay = getPlatformDisplayEXT(platform, nativeDisplay, nullptr);
+
+    if (m_eglDisplay == EGL_NO_DISPLAY)
     {
-      m_eglDisplay = getPlatformDisplayEXT(EGL_PLATFORM_GBM_KHR, (EGLNativeDisplayType)display, NULL);
+      CEGLUtils::LogError("Failed to get EGL platform display");
+      return false;
```

Comment:
If eglGetPlatformDisplayEXT fails, do we want to fall back to eglGetDisplay?

Reply:
yes, that's what we've been doing

Reply:
Hm I wonder, if that failed does that also mean we cannot use `eglCreatePlatformWindowSurfaceEXT` afterwards?

Reply:
To simplify this: Would you think it'd be OK to drop the fallback on failure and do the fallback only if EGL_EXT_platform_base and the specific platform extension are unsupported?

Reply:
Yea that's probably a good idea